File: go/array/648_replace_words.go
// Time Complexity: O(n * m)
// Space Complexity: O(d + n)
package main

import "strings"

func replaceWords(dictionary []string, sentence string) string {
	rootSet := make(map[string]bool)
	for _, root := range dictionary {
		rootSet[root] = true
	}

	words := strings.Fields(sentence)

	for i, word := range words {
		for j := 1; j <= len(word); j++ {
			prefix := word[:j]

			if rootSet[prefix] {
				words[i] = prefix
				break
			}
		}
	}

	return strings.Join(words, " ")
}

File: python/array/648_replace_words.py
# Time Complexity: O(n * m)
# Space Complexity: O(d + n)

from typing import List

class Solution:
    def replaceWords(self, dictionary: List[str], sentence: str) -> str:
        root_set = set(dictionary)
        words = sentence.split()

        for i in range (len(words)):
            for j in range(1, len(words[i]) + 1):
                prefix = words[i][:j]
                if prefix in root_set:
                    words[i] = prefix
                    break

        
        return " ".join(words)